module-bluetooth/Bluetooth/interface/profiles/GAP/GAP.cpp

```diff
@@ -94,48 +109,28 @@ namespace bluetooth
         return gap_inquiry_start(inquiryIntervalSeconds);
     }
 
-    auto GAP::remoteNameToFetch() -> bool
-    {
-        auto result = std::find_if(std::begin(devices), std::end(devices), [](Devicei &device) {
-            return device.state == REMOTE_NAME_REQUEST;
-        });
-
-        return result != std::end(devices);
-    }
-
-    void GAP::fetchRemoteName()
-    {
-        for (auto &device : bluetooth::GAP::devices) {
-            if (device.state == REMOTE_NAME_REQUEST) {
-                device.state = REMOTE_NAME_INQUIRED;
-                LOG_INFO("Get remote name...");
-                gap_remote_name_request(device.address, device.pageScanRepetitionMode, device.clockOffset | 0x8000);
-                return;
-            }
-        }
-    }
-
     void GAP::continueScanning()
     {
-        if (remoteNameToFetch()) {
-            fetchRemoteName();
+        if (const auto &it = devices().find(REMOTE_NAME_REQUEST); it != devices().end()) {
+            LOG_INFO("Get remote name for %s", it->name.data());
+            it->state = REMOTE_NAME_INQUIRED;
+            gap_remote_name_request(it->address, it->pageScanRepetitionMode, it->clockOffset | 0x8000);
             return;
         }
-        startScan();
+        if ( gap::state == gap::state::scan_on) {
+            startScan();
+        }
     }
+
     auto GAP::updateDeviceName(std::uint8_t *packet, bd_addr_t &addr) -> bool
     {
         reverse_bd_addr(&packet[3], addr);
-        auto index = getDeviceIndexForAddress(devices, addr);
-        if (index >= 0) {
-            if (packet[2] == 0) {
-                devices[index].state = REMOTE_NAME_FETCHED;
-                strcpy(devices[index].name.data(), reinterpret_cast<const char *>(&packet[9]));
-                return true;
-            }
-            else {
-                LOG_INFO("Failed to get name: page timeout");
+        if (auto it = devices().find(addr); it != devices().end()) {
+            it->state = packet[2] ? REMOTE_NAME_FAILURE : REMOTE_NAME_FETCHED;
+            if ( it->state != REMOTE_NAME_FAILURE ) {
```

Comment:
while at it, consider naming 3 and 2.

Reply:
tbh this is taken straight from packet definition - this should be rather addressed in Bluetooth stack, not in our code. in most scenarios error code is on 2'nd field, bt for some it's not :)